File: backend/services/subtasks.py
"""The lines under a task: reading them, and the two rules about what one may be.

A line is a `blocks` row with a `parent_id` — `migrations/008_subtasks.sql` says why — so the
writes that change one (tick it, rename it, remove it) are the writes the API already has, and
nothing here duplicates them. What is here is everything about the *relationship*: reading the
lines of a set of tasks in one query, giving a new line its place, and refusing the two shapes
that would make a checklist into something else.

Four decisions are worth reading before the code:

**One level.** A line of a line is refused, with a sentence. SQLite would hold one happily; a
checklist under a checklist is a document, and this app is a day. So there is exactly one
`parent_id` deep, and `parent_row` is the only place that decides it.

**A line has no day of its own.** `day` and `start_min` stay NULL: the line is inside its task,
so it is read with its task and never by day. That is what keeps the day's own query, the week's
counts, the inbox and the rollover from having to know lines exist — and it is why the inbox
query says `parent_id IS NULL` beside its `day IS NULL`: an inbox is a list of tasks you can give
a time to.

**A line is created as a line.** Nothing re-parents one, so `BlockPatch` has no `parent_id` and
there is no route that makes a task out of a line. Moving a line between tasks is a thing nobody
has asked for, and every way of doing it is a way to lose the ticks on it.

**Sorting is stored, not implied.** A line carries the `sort_order` it was written at, because an
export selects rows and an import re-inserts them: the order a person wrote down belongs in the
file rather than in whatever rowid SQLite hands out. Ties (two lines written at once) fall back
to the rowid, which is the order they were inserted in.
"""

from __future__ import annotations

from typing import Sequence

from fastapi import HTTPException

from services.blocks import row_to_dict


def lines_of(conn, parents: Sequence[str]) -> dict[str, list[dict]]:
    """The lines of these tasks, in the order they were written, keyed by task id.

    One query for the whole set rather than one per task, because the day view asks about every
    block on the day at once and a query per block is how a page of twenty tasks becomes forty
    round trips to SQLite.
    """
    ids = [p for p in parents if p]
    if not ids:
        return {}
    marks = ", ".join("?" for _ in ids)
    rows = conn.execute(
        f"SELECT * FROM blocks WHERE parent_id IN ({marks}) ORDER BY sort_order, rowid",
        ids,
    ).fetchall()

    found: dict[str, list[dict]] = {}
    for row in rows:
        found.setdefault(row["parent_id"], []).append(row_to_dict(row))
    return found


def attach(conn, blocks: list[dict]) -> list[dict]:
    """Put each task's checklist on it, and hand the list back.

    Every block gets the key, an empty list included. A client that has to test for the key's
    absence is a client that will one day treat "no lines" and "this shape is new to me" as the
    same thing.
    """
    lines = lines_of(conn, [b["id"] for b in blocks])
    for block in blocks:
        block["subtasks"] = lines.get(block["id"], [])
    return blocks


def next_order(conn, parent_id: str) -> int:
    """Where a new line goes: after the ones already there."""
    row = conn.execute(
        "SELECT COALESCE(MAX(sort_order), -1) AS n FROM blocks WHERE parent_id = ?", (parent_id,)
    ).fetchone()
    return int(row["n"]) + 1


def parent_row(conn, parent_id: str):
    """The task a new line is being added to, or a refusal naming what is wrong.

    Two refusals, because they are two different mistakes. A task that is not there is a 404;
    a task that is itself a line is a 400, and the sentence says which rule it broke rather than
    leaving a client to guess that nesting is one deep.
    """
    row = conn.execute("SELECT * FROM blocks WHERE id = ?", (parent_id,)).fetchone()
    if row is None:
        raise HTTPException(404, "no such block")
    if row["parent_id"] is not None:
        raise HTTPException(400, "a checklist line cannot hold lines of its own")
    return row


def check_patch(current: dict, fields: dict, unschedule: bool = False) -> None:
    """A line is inside its task: it is not scheduled, moved to a day, or sent to the inbox.

    Checked as a whole rather than field by field at the point of use, because the pair travels
    together: `unschedule` is two columns, and a line given one of the three would otherwise end
    up half scheduled — which the table's own CHECK refuses with a 500 rather than a sentence.
    """
    if current.get("parent_id") is None:
        return
    if unschedule or "day" in fields or "start_min" in fields:
        raise HTTPException(400, "a checklist line is inside its task, not on the day")

File: frontend/src/subtasks.js
/** A task's checklist: the steps it holds, and where a write about one of them goes.
 *
 *  A pure module, like `templates.js` and `rollover.js`, because everything in it is worth
 *  asserting without a browser. The server owns the order — a line carries the `sort_order` it
 *  was written at — and owns every refusal. What lives here is the local half: which list of
 *  steps is on screen, which route changes one, and the list edits made between saves.
 *
 *  Three different things can hold a checklist, and they are three different shapes, which is
 *  the whole reason this file exists. A *task* holds lines that are blocks of their own, so
 *  ticking one is a PATCH to that row. A *routine* holds definitions: the steps belong to the
 *  rule and are drawn on every day of it, while ticking one is a fact about a single day that
 *  goes to that day's occurrence — the rule's own editor has no day to tick. A *template item*
 *  holds steps that are part of the item list, saved with it. `whereSteps` is the one place that
 *  decision is made.
 *
 *  Nothing here counts anything. The steps are on the screen under their task, which is the
 *  progress; a "2 of 3" beside the title would be a score, and this app does not keep those.
 */

import { isOccurrence } from './routines.js'

/** The steps a task, a rule or a template item holds. Always an array.
 *
 *  A client that has to test for the key's absence is a client that will one day read "no
 *  steps" and "this shape is new to me" as the same thing — and the API sends the key on every
 *  block for exactly that reason. */
export const stepsOf = (holder) => (Array.isArray(holder?.subtasks) ? holder.subtasks : [])

/** Whether anything draws at all. An empty checklist renders nothing, which is what keeps a
 *  task with no steps looking exactly as it did before any of this existed. */
export const hasSteps = (holder) => stepsOf(holder).length > 0

/** What a template step starts as when you add one. A word rather than an empty box, so the
 *  list is never in a state the API refuses — the same reasoning as `NEW_ITEM`. */
export const NEW_STEP = { title: 'New step' }

/** Add a step at the end (a template item's draft list: positional, no ids yet). */
export const withStep = (steps = [], step = NEW_STEP) => [...steps, { ...step }]

/** Take the step at `index` out. */
export const withoutStep = (steps, index) => steps.filter((_, i) => i !== index)

/** Rename the step at `index`, leaving the rest of the list — and its order — alone. */
export const changeStep = (steps, index, title) =>
  steps.map((step, i) => (i === index ? { ...step, title } : step))

/** The steps as the template's own body sends them: names, in order, and nothing else. Ids and
 *  `sort_order` are the server's to assign, and a step has no hour or colour of its own. */
export const payloadSteps = (steps = []) =>
  steps.map((step) => ({ title: String(step?.title ?? '').trim() }))

/** Whether every step in the list has a name. One blank line is a body the API refuses whole,
 *  so the panel holds the save rather than sending it — the same rule the items follow. */
export const namesEveryStep = (steps = []) =>
  steps.every((step) => String(step?.title ?? '').trim().length > 0)

/**
 * Where a write about one of these steps goes.
 *
 * Returns `null` when there is no checklist to speak of, otherwise a descriptor naming the
 * routes: `{kind: 'block', blockId}` for a task, `{kind: 'occurrence', routineId, day}` for a
 * day of a routine, `{kind: 'routine', routineId, day: null}` for the rule itself.
 *
 * A rule has no day, so nothing about it can be ticked: its steps are definitions, and the tick
 * belongs to a morning. That is what `tickable` says, and the editor draws no box when it is
 * false rather than drawing one that writes nothing.
 */
export function whereSteps(subject) {
  if (!subject) return null
  if (subject.kind === 'routine') {
    const rule = subject.routine
    if (!rule) return null
    return { kind: 'routine', routineId: rule.id, day: null, tickable: false }
  }
  const block = subject.block
  if (!block) return null
  if (isOccurrence(block)) {
    return {
      kind: 'occurrence',
      routineId: block.routine_id,
      day: block.occurrence_day ?? block.day,
      tickable: true,
    }
  }
  return { kind: 'block', blockId: block.id, day: block.day ?? null, tickable: true }
}
